lib/ASTUtils.cc

```diff
@@ -10,114 +9,101 @@
+        return TT->getAsTagDecl();
+    }
+    return nullptr;
+}
 
-        auto ND = clang::dyn_cast<clang::NamedDecl>(D);
-        if (!ND) {
-            return false;
-        }
+bool isBaseTypeAnonymous(const clang::Type *T) {
+    auto TD = getBaseTypeTagDeclOrNull(T);
+    auto ND = clang::dyn_cast_or_null<clang::NamedDecl>(TD);
 
-        return ND->getIdentifier() == nullptr || ND->getName().empty();
-    });
+    return ND && (!ND->getIdentifier() || ND->getName().empty());
 }
 
-bool hasLocalType(const clang::Type *T, clang::ASTContext &Ctx) {
-    return isInType(T, Ctx, [](const clang::Type *T) {
-        if (!T) {
-            return false;
-        }
+bool isBaseTypeLocal(const clang::Type *T) {
+    auto TD = getBaseTypeTagDeclOrNull(T);
+    auto DCtx = TD ? TD->getDeclContext() : nullptr;
 
-        auto D = getTypeDeclOrNull(T);
-        if (!D) {
-            return false;
-        }
+    return DCtx && !DCtx->isTranslationUnit();
+}
 
-        auto DCtx = D->getDeclContext();
+bool isBaseTypeDefinedAfter(const clang::Type *T, clang::ASTContext &Ctx,
+                            clang::SourceLocation L) {
+    auto &SM = Ctx.getSourceManager();
+    auto TD = getBaseTypeTagDeclOrNull(T);
+    auto DLoc = TD ? TD->getLocation() : clang::SourceLocation();
+    auto DFLoc = DLoc.isValid() ? SM.getFileLoc(DLoc) : clang::SourceLocation();
 
-        return !DCtx->isTranslationUnit();
-    });
+    return DFLoc.isValid() && SM.isBeforeInTranslationUnit(L, DFLoc);
 }
 
-bool hasTypeDefinedAfter(const clang::Type *T, clang::ASTContext &Ctx,
-                         clang::SourceLocation L, bool IgnoreTypedefs) {
+bool isAnyDeusgaredTypeDefinedAfter(const clang::Type *T,
+                                    clang::ASTContext &Ctx,
+                                    clang::SourceLocation L) {
     auto &SM = Ctx.getSourceManager();
 
-    if (!T) {
-        return false;
-    }
-
-    if (!IgnoreTypedefs) {
-        auto TD = clang::dyn_cast<clang::TypedefType>(T);
-        if (!TD) {
-            auto ET = clang::dyn_cast<clang::ElaboratedType>(T);
-            if (ET) {
-                TD = clang::dyn_cast<clang::TypedefType>(ET->getNamedType());
-            }
-        }
-        if (TD) {
-            auto D = TD->getDecl();
-            auto DLoc = D->getLocation();
-            if (DLoc.isValid()) {
-                auto DFLoc = SM.getFileLoc(DLoc);
-                if (DFLoc.isValid()) {
-                    return SM.isBeforeInTranslationUnit(L, DFLoc);
-                }
+    while (T) {
+        if (T->isArrayType()) {
+            T = T->getBaseElementTypeUnsafe();
+        } else if (T->isAnyPointerType()) {
+            T = T->getPointeeType().getTypePtrOrNull();
+        } else if (auto ET = clang::dyn_cast<clang::ElaboratedType>(T)) {
+            T = ET->desugar().getTypePtrOrNull();
+        } else if (auto TT = clang::dyn_cast_or_null<clang::TagType>(T)) {
```

Comment:
`dyn_cast` suffices here; I used `dyn_cast_or_null` because I had copied and pasted it from another function definition. I won't change the existing uses to `dyn_cast_or_present` but will use that going first, thanks for finding out that `dyn_cast_or_null` is deprecated.